third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp

```diff
@@ -126,10 +126,146 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
+      isa<arith::SubIOp>(reduceOp) || isa<arith::XOrIOp>(reduceOp)) {
+    return reduceOp;
+  }
+
+  return std::nullopt;
+}
+
 bool TargetInfo::warpReduce(RewriterBase &rewriter, Location loc,
                             SmallVector<Value> &acc, triton::ReduceOp op,
                             unsigned numLaneToReduce,
                             unsigned interleave) const {
+  if (numLaneToReduce != 64)
+    return false;
+
+  if (auto optionalReduxOp = getReduxOpForDPP(op)) {
+    auto reduxOp = *optionalReduxOp;
+
+    auto createDppReduxOp = [&rewriter, &loc, &op,
+                             &reduxOp](Type valType, Value &src, int dppCtrl,
+                                       int rowMask, int bankMask,
+                                       bool boundCtrl) -> Value {
+      unsigned bits = valType.getIntOrFloatBitWidth();
+
+      Type actualType = valType;
+      if (!valType.isF32()) {
+        if (!valType.isIntOrIndex()) {
+          src = bitcast(src, int_ty(bits));
+          actualType = int_ty(bits);
+        }
+
+        if (bits < 32) {
+          src = sext(i32_ty, src);
+          actualType = i32_ty;
+        }
+      }
+
+      auto dppResult = rewriter
+                           .create<ROCDL::DPPUpdateOp>(
+                               loc, actualType, src, src,
+                               mlir::IntegerAttr::get(i32_ty, dppCtrl),
+                               mlir::IntegerAttr::get(i32_ty, rowMask),
+                               mlir::IntegerAttr::get(i32_ty, bankMask),
+                               mlir::BoolAttr::get(op.getContext(), boundCtrl))
+                           .getRes();
+
+      if (!valType.isF32()) {
+        if (bits < 32) {
+          src = trunc(int_ty(bits), src);
+          dppResult = trunc(int_ty(bits), dppResult);
+        }
+
+        if (!valType.isIntOrIndex()) {
+          src = bitcast(src, valType);
+          dppResult = bitcast(dppResult, valType);
+        }
+      }
+
+      IRMapping mapping;
+      mapping.map(reduxOp->getOperand(0), src);
+      mapping.map(reduxOp->getOperand(1), dppResult);
+      return rewriter.clone(*reduxOp, mapping)->getResult(0);
+    };
+
+    for (int i = 0; i < acc.size(); i++) {
+      Value buf;
+      auto valType = acc[i].getType();
+
+      // row_shr:8 bank_mask:0xc
+      buf = createDppReduxOp(valType, acc[i], 8 + 0x110, 0xF, 0xC, true);
+
+      // row_shr:4 bank_mask:0x8
+      buf = createDppReduxOp(valType, buf, 4 + 0x110, 0xF, 0x8, true);
+
+      // row_shr:2
+      buf = createDppReduxOp(valType, buf, 2 + 0x110, 0xF, 0xF, true);
+
+      // row_shr:1
+      buf = createDppReduxOp(valType, buf, 1 + 0x110, 0xF, 0xF, true);
+
+      // row_bcast:15 row_mask:0xa
+      buf = createDppReduxOp(valType, buf, 0x142, 0xA, 0xF, true);
+
+      // row_bcast:31
+      buf = createDppReduxOp(valType, buf, 0x143, 0xF, 0xF, true);
+
+      Type actualType = valType;
+      unsigned bits = valType.getIntOrFloatBitWidth();
```

Comment:
Similarly here. Add code block comment. It's also useful to use some `static` utility functions like `castToIntGE16` or something to be clearer.